Tools/CISupport/ews-build/steps.py

```diff
@@ -7302,43 +7310,170 @@ def run(self):
+            if successful_filter and self.filtered_changes and unexpected_results_after_filter:  # If there are unexpected results
+                yield self._addToLog('stdio', f'\nSuccessfully filtered results! Updating unexpected_results.json on disk.\n')
+                yield self.write_unexpected_results_file()
+            if not successful_filter:
+                yield self._addToLog('stdio', f'\nFailed to consult results database. Falling back to tip-of-tree...\n')
+                # If results db failed, rebuild without changes to verify causation
+                self.build.addStepsAfterCurrentStep([ValidateChange(verifyBugClosed=False, addURLs=False), RevertAppliedChanges(exclude=['new*', 'scan-build-output*']), ScanBuildWithoutChange()])
+                self.createResultMessage()
+                return defer.returnValue(rc)
 
-        unexpected_results = self.getProperty('unexpected_failing_files', 0) or self.getProperty('unexpected_new_issues', 0) or self.getProperty('unexpected_passing_files', 0)
-        if self.expectations and unexpected_results:
-            # If there are unexpected results, rebuild without changes to verify causation
-            self.build.addStepsAfterCurrentStep([ValidateChange(verifyBugClosed=False, addURLs=False), RevertAppliedChanges(exclude=['new*', 'scan-build-output*']), ScanBuildWithoutChange()])
+        # Only save the results if there are unexpected results
+        if self.expectations:
+            if not unexpected_results_after_filter:
+                yield self._addToLog('stdio', f'Found no unexpected results after filtering through results database!\n')
+            else:
+                yield self._addToLog('stdio', f'Found unexpected results after filtering through results database!\n')
+                steps_to_add = [DownloadUnexpectedResultsFromMaster(), DeleteStaticAnalyzerResults(results_dir='StaticAnalyzerUnexpectedRegressions')] if self.filtered_changes else []
+                steps_to_add += [GenerateSaferCPPResultsIndex(), DeleteStaticAnalyzerResults(), ArchiveStaticAnalyzerResults(), UploadStaticAnalyzerResults(), ExtractStaticAnalyzerTestResults(), DisplaySaferCPPResults()]
+                self.build.addStepsAfterCurrentStep(steps_to_add)
         elif unexpected_results:
-            # Only save the results if there are failures and it is not the first run
             self.build.addStepsAfterCurrentStep([ArchiveStaticAnalyzerResults(), UploadStaticAnalyzerResults(), ExtractStaticAnalyzerTestResults(), DisplaySaferCPPResults()])
+
+        self.createResultMessage()
+
         return defer.returnValue(rc)
 
-    def createResultMessage(self):
+    @defer.inlineCallbacks
+    def write_unexpected_results_file(self):
+        resultDirectory = f"public_html/results/{self.getProperty('buildername')}"
```

Comment:
Should we be concerned about this path being written / accessed by multiple `macOS-Safer-CPP-Checks-EWS` builds simultaneously?

Reply:
Ah. Great point - let me modify this function so it creates a directory for each build.

Reply:
Here's a successful run after changing the directory: https://ews-build.webkit-dev.org/#/builders/218/builds/1741